src/index.js

```diff
@@ -214,53 +203,74 @@ export default function undoable (reducer, rawConfig = {}) {
     jumpType: rawConfig.jumpType || ActionTypes.JUMP,
     clearHistoryType: rawConfig.clearHistoryType || ActionTypes.CLEAR_HISTORY
   }
-  config.history = rawConfig.initialHistory || createHistory(config.initialState || reducer(undefined, {}))
 
-  return (state = config.history, action = {}) => {
+  return (state, action = {}) => {
     debugStart(action, state)
+
+    let history
+    if (!config.history) {
+      debug('create history on init')
+
+      if (state === undefined) {
+        config.history = createHistory(reducer(state, {}))
+      } else if (isHistory(state)) {
+        config.history = state
+      } else {
+        config.history = createHistory(state)
+      }
+
+      history = config.history
+    } else if (state === undefined) {
+      // If reducer is called with undefined, use our saved history as the state
+      // since that was the result of calling the reducer the initially
+      history = config.history
+    } else {
+      history = state
+    }
+
     let res
     switch (action.type) {
       case undefined:
-        return state
+        return history
 
       case config.undoType:
-        res = undo(state)
+        res = undo(history)
```

Comment:
state is alway a full history object at this point, so lets name it like one for clarity (and because wen can't use the param name `state` anyway.. )